crates/sui-bcs-json/src/lib.rs

```diff
@@ -0,0 +1,841 @@
+        }
+        "u128" => {
+            let bcs: u128 = bcs::from_reader_no_check(data)?;
+            Ok(serde_json::to_value(bcs.to_string())?)
+        }
+        "u256" => {
+            let bcs: U256 = bcs::from_reader_no_check(data)?;
+            Ok(serde_json::to_value(bcs.to_string())?)
+        }
+        other => bail!("Error: expected u8,u16,u32,u64,u128, or u256, but got {other}."),
+    }
+}
+
+/// Return the inner type if the outer type is a vector
+fn is_vector(typename: &str) -> Option<&str> {
+    typename.strip_prefix("vector<")?.strip_suffix(">")
+}
+
+//**************************************************************************************************
+// To BCS helper functions
+//**************************************************************************************************
+
+/// Return a BCS array from the given JSON object.
+/// The function returns an error if the nested vectors' values do not have consistent types.
+fn internal_bcs_from_json_vector(val: &Value, output: &mut Vec<u8>, typename: &str) -> Result<()> {
+    match val {
+        Array(arr) if typename.starts_with("vector") => {
+            // before we the type, we need to strip the prefix type
+            let typename = typename
+                .strip_prefix("vector<")
+                .ok_or(anyhow!("Cannot strip vector prefix"))?
+                .strip_suffix(">")
+                .ok_or(anyhow!("Cannot strip vector suffix"))?;
+            // bcs standard requires the length of the arrays that are being decoded
+            output.insert(output.len(), arr.len() as u8);
+            for i in arr {
+                internal_bcs_from_json_vector(i, output, typename)?;
+            }
+        }
+        other => output.extend(
+            internal_bcs_from_json(typename, &serde_json::to_string(other)?).map_err(|e| {
+                anyhow!("Expected a (nested) vector of {typename}, and got {e} instead.")
+            })?,
+        ),
+    }
+    Ok(())
+}
+
+fn internal_bcs_from_json(typename: &str, data: &str) -> Result<Vec<u8>> {
+    let type_ = if typename.starts_with("vector") {
+        "vector"
+    } else {
+        typename
+    };
+    let bcs_bytes = match type_ {
+        "u8" | "u16" | "u32" | "u64" | "u128" | "u256" => string_integer_to_bcs(typename, data),
+        "bool" => to_bcs::<bool>(data),
+        "vector" => to_bcs_from_vector(typename, data),
+        "Argument" => to_bcs::<Argument>(data),
+        "CallArg" => to_bcs::<CallArg>(data),
+        "ObjectDigest" => to_bcs::<ObjectDigest>(data),
+        "ObjectID" => to_bcs::<ObjectID>(data),
+        "ProgrammableTransaction" => to_bcs::<ProgrammableTransaction>(data),
+        "SequenceNumber" => to_bcs::<SequenceNumber>(data),
+        "SuiAddress" => to_bcs::<SuiAddress>(data),
+        "String" => to_bcs::<String>(data),
+        "TransactionData" => to_bcs::<TransactionData>(data),
+        "TransactionKind" => to_bcs::<TransactionKind>(data),
```

Comment:
I'm not sure i fully understand how this library is intended to be used. Given the stated goal in the README i expect this is intended as a mechanism for other languages which don't have bcs support (serde-reflection provides bcs support for a number of languages though) the ability to write and craft a transaction in JSON and then have this tool take that and then write out its canonical BCS format. Given that i'm not sure why this tool would support all of these types and not just the high level "Transaction" type or other top-level sui data types. I would not expect anyone to create a transaction by going back and forth multiple times into this library but instead would expect it to be a more one-shot style where you have a fully formed txn.

Reply:
> write and craft a transaction in JSON and then have this tool take that and then write out its canonical BCS format.

This is mostly true, but there are some edge cases that require the introduction of finer grained types:

- Serialization of `Pure` transaction arguments, which requires the the ability serialise primitive types and vectors of those.
- The ability to ser/de arbitrary Move types, which would be composed of those primitives -- we briefly entertained the idea of supporting a DSL for describing such types so that you could call into this library once with a DSL describing the schema (the extensions we are planning for Object Display require a similar DSL which we were thinking of borrowing) -- but I thought it would be less jarring to offer an API that allows the client to chain calls into the serializer/deserializer for these cases.
- I suggested that @stefan-mysten look through the types that the TS SDK serializes and deserializes to BCS to get an indication of the kinds of calls an SDK in another language might make.

Some of the types that are mentioned here may not fit into any of these cases and we can get rid of those though (`Argument`, `CallArg`, `ProgrammableTransaction` come to mind).

Reply:
Thanks both for your comments. I have not removed those types before asking for review, which I understand now that it created a lot of confusion. Here's what TS BCS supports: https://sui-typescript-docs.vercel.app/typescript/bcs and here https://sui-typescript-docs.vercel.app/bcs#description

Reply:
Note that there is a slight difference between the list of supported types by TS BCS that you've linked here and the list that I meant.  We only need to support the "root" types that actually get passed to `bcs.ser` or `bcs.se` in the TS SDK.  The list above includes the names of all intermediate types as well, which we don't need here, because that mapping is handled within Rust by serde derivations.

Reply:
Right!